### PR TITLE
suport ot activation

### PR DESCRIPTION
**Pull request by ascrutae**
Ref: #270

### CODE REVIEW

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SkywalkingSpanActivation.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        return "org.skywalking.apm.toolkit.opentracing.SkywalkingSpan";
+    }
+
+    @Override
+    protected ConstructorInterceptPoint[] getConstructorsInterceptPoints() {
+        return new ConstructorInterceptPoint[] {
+            new ConstructorInterceptPoint() {
+                @Override
+                public ElementMatcher<MethodDescription> getConstructorMatcher() {
+                    return takesArgument(0, SkywalkingSpanBuilder.class);
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SkywalkingSpanActivation.java]: Potential class loader issue.

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SkywalkingSpanActivation.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+                @Override
+                public String getConstructorInterceptor() {
+                    return "org.skywalking.apm.toolkit.activation.opentracing.span.ConstructorWithSpanBuilderInterceptor";
+                }
+            },
+            new ConstructorInterceptPoint() {
+                @Override
+                public ElementMatcher<MethodDescription> getConstructorMatcher() {
+                    return takesArgument(0, SkywalkingTracer.class);
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SkywalkingSpanActivation.java]: Potential class loader issue.

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SpanFinishInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, String methodName, Object[] allArguments,
+        Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, String methodName, Object[] allArguments,
+        Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        AbstractTracingSpan abstractSpan = (AbstractTracingSpan)objInst.getSkyWalkingDynamicField();
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SpanFinishInterceptor.java]: You can't guarantee to get `AbstractTracingSpan`, if this trace isn't sampled.

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/ConstructorWithSpanBuilderInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        if (spanBuilder.isEntry()) {
+            ContextManager.createEntrySpan(spanBuilder.getOperationName(), null);
+        } else if (spanBuilder.isExit() && (!StringUtil.isEmpty(spanBuilder.getPeer()))) {
+            ContextManager.createExitSpan(spanBuilder.getOperationName(),
+                new ContextCarrier(), buildRemotePeer(spanBuilder));
+        } else {
+            ContextManager.createLocalSpan(spanBuilder.getOperationName());
+        }
+
+        AbstractSpan span = ContextManager.activeSpan();
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/ConstructorWithSpanBuilderInterceptor.java]: No need to use `activeSpan()` method, `create*Span` method returns the `AbstractSpan`.

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SpanLogInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, String methodName, Object[] allArguments,
+        Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, String methodName, Object[] allArguments,
+        Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        AbstractTracingSpan abstractSpan = (AbstractTracingSpan)objInst.getSkyWalkingDynamicField();
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SpanLogInterceptor.java]: You can't guarantee to get AbstractTracingSpan, if this trace isn't sampled.

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SpanSetOperationNameInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, String methodName, Object[] allArguments,
+        Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
+
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, String methodName, Object[] allArguments,
+        Class<?>[] argumentsTypes, Object ret) throws Throwable {
+        AbstractTracingSpan tracingSpan = (AbstractTracingSpan)objInst.getSkyWalkingDynamicField();
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-opentracing-activation/src/main/java/org/skywalking/apm/toolkit/activation/opentracing/span/SpanSetOperationNameInterceptor.java]: You can't guarantee to get AbstractTracingSpan, if this trace isn't sampled.

**Comment by ascrutae** [comment]
Here is the trace segment with `contiunation`
```
{
    "traceSegmentId": "S.1499700553732.144823023.84411.20.1",
    "refs": [
        {
            "type": "CROSS_THREAD",
            "traceSegmentId": "S.1499700545721.144823023.84411.1.1",
            "spanId": 2,
            "applicationInstanceId": 0,
            "peerId": 0,
            "operationId": 0
        }
    ],
    "spans": [
        {
            "spanId": 0,
            "parentSpanId": -1,
            "operationName": "AsyncThread/queryData",
            "operationId": 0,
            "startTime": 1499700553732,
            "endTime": 1499700554737,
            "errorOccurred": false,
            "componentId": 0
        }
    ],
    "relatedGlobalTraces": {
        "relatedGlobalTraces": [
            {
                "id": "T.1499700545729.144823023.84411.1.2"
            }
        ]
    },
    "ignore": false
}
```

Here is the trace segment of generate `contiunation`
```
{
    "traceSegmentId": "S.1499700545721.144823023.84411.1.1",
    "spans": [
        {
            "spanId": 2,
            "parentSpanId": 1,
            "tags": [
                {
                    "key": "name",
                    "value": "Test"
                }
            ],
            "operationName": "AsyncDao/updateDataAsync",
            "operationId": 0,
            "startTime": 1499700545768,
            "endTime": 1499700551560,
            "errorOccurred": false,
            "componentId": 0
        },
        {
            "spanId": 1,
            "parentSpanId": 0,
            "tags": [
                {
                    "key": "name",
                    "value": "Test"
                }
            ],
            "operationName": "AsyncService/updataDataAsync",
            "operationId": 0,
            "startTime": 1499700545767,
            "endTime": 1499700552992,
            "errorOccurred": false,
            "componentId": 0
        },
        {
            "spanId": 0,
            "parentSpanId": -1,
            "operationName": "AsyncProject/Main",
            "operationId": 0,
            "startTime": 1499700545762,
            "endTime": 1499700554996,
            "errorOccurred": false,
            "componentId": 0
        }
    ],
    "relatedGlobalTraces": {
        "relatedGlobalTraces": [
            {
                "id": "T.1499700545729.144823023.84411.1.2"
            }
        ]
    },
    "ignore": false
}
```

File: apm-sniffer/apm-agent-core/src/main/java/org/skywalking/apm/agent/core/context/trace/AbstractTracingSpan.java [hunk 7]
```diff
@@ -147,7 +147,7 @@ public AbstractSpan errorOccurred() {
      * @param operationName
      * @return span instance, for chaining.
      */
-    public AbstractTracingSpan setOperationName(String operationName) {
+    public AbstractSpan setOperationName(String operationName) {
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/skywalking/apm/agent/core/context/trace/AbstractTracingSpan.java]: Why change this return type?

**Pull request merged**